crates/iceberg/src/spec/table_metadata_builder.rs

```diff
@@ -0,0 +1,2070 @@
+    pub fn add_snapshot(mut self, snapshot: Snapshot) -> Result<Self> {
+        if self
+            .metadata
+            .snapshots
+            .contains_key(&snapshot.snapshot_id())
+        {
+            return Err(Error::new(
+                ErrorKind::DataInvalid,
+                format!("Snapshot already exists for: '{}'", snapshot.snapshot_id()),
+            ));
+        }
+
+        if self.metadata.format_version != FormatVersion::V1
+            && snapshot.sequence_number() <= self.metadata.last_sequence_number
+            && snapshot.parent_snapshot_id().is_some()
+        {
+            return Err(Error::new(
+                ErrorKind::DataInvalid,
+                format!(
+                    "Cannot add snapshot with sequence number {} older than last sequence number {}",
+                    snapshot.sequence_number(),
+                    self.metadata.last_sequence_number
+                )
+            ));
+        }
+
+        if let Some(last) = self.metadata.snapshot_log.last() {
+            // commits can happen concurrently from different machines.
+            // A tolerance helps us avoid failure for small clock skew
+            if snapshot.timestamp_ms() - last.timestamp_ms < -ONE_MINUTE_MS {
+                return Err(Error::new(
+                    ErrorKind::DataInvalid,
+                    format!(
+                        "Invalid snapshot timestamp {}: before last snapshot timestamp {}",
+                        snapshot.timestamp_ms(),
+                        last.timestamp_ms
+                    ),
+                ));
+            }
+        }
+
+        if snapshot.timestamp_ms() - self.metadata.last_updated_ms < -ONE_MINUTE_MS {
+            return Err(Error::new(
+                ErrorKind::DataInvalid,
+                format!(
+                    "Invalid snapshot timestamp {}: before last updated timestamp {}",
+                    snapshot.timestamp_ms(),
+                    self.metadata.last_updated_ms
+                ),
+            ));
+        }
+
+        // Mutation happens in next line - must be infallible from here
+        self.changes.push(TableUpdate::AddSnapshot {
+            snapshot: snapshot.clone(),
+        });
+
+        self.metadata.last_updated_ms = snapshot.timestamp_ms();
+        self.metadata.last_sequence_number = snapshot.sequence_number();
+        self.metadata
+            .snapshots
+            .insert(snapshot.snapshot_id(), snapshot.into());
+
+        Ok(self)
+    }
+
+    /// Append a snapshot to the specified branch.
+    /// If branch is not specified, the snapshot is appended to the main branch.
+    /// The `ref` must already exist. Retention settings from the `ref` are re-used.
+    ///
+    /// # Errors
+    /// - The ref is unknown.
+    /// - Any of the preconditions of `self.add_snapshot` are not met.
+    pub fn append_snapshot(self, snapshot: Snapshot, ref_name: Option<&str>) -> Result<Self> {
```

Comment:
A few thoughts here:
1. The `append_snapshot` method has been around for quite some time now: https://github.com/apache/iceberg-rust/blob/50345196c87b00badc1a6490aef284e84f4c3e9a/crates/iceberg/src/spec/table_metadata.rs#L337 - not sure how many people use it
2. This method does `append` the snapshot to the `snapshot_log` if `ref_name` is `None` or `'main'`. I think this is where the name originated from.
3. There is nothing technical stopping us from not specifying a `tag` instead of a branch - hence I chose to use `ref_name` as an argument name. We use the same argument name also for `set_ref`

I am also fine with going ahead with renaming. Just wanted us to consider those points as well.
Let me know if you want to go ahead @liurenjie1024

Reply:
I kept the old method but I am not sure how to deprecate it.
I saw that in other crates we are using log::warn - but we currently don't depend on the log crate. How should we deprecate the method?

I added. the `set_branch_snapshot` which now mostly follows java logic. It's also taking a `branch` argument now and fails if it gets a tag - just like java. Two key differences:
1. I don't check the sequence number, as its already checked in `append_snapshot`
2. If the branch does not exist, I fail instead of creating it. I don't like creating a branch implicitly here, as there is no option to configure Retention for that branch. I see the convenience in "just create a branch", but Retention is important and I am not sure if the builder as the foundation should already be opinionated about it. Let me know if you see it differently, then I would create the branch if it doesn't exist.

Reply:
Ok, deprecated the `TableMetadata` method and also removed the `append_snapshot` method from the builder.
`set_snapshot_ref` is now the new alternative in the builder. 
Let me know what you think about deviating to java as describe in my last comment.

Reply:
I thought about this method again yesterday. While I don't like having no way in the API to set default retention, most users actually don't specify retention explicitly. I believe it's better to set retention as part of `iceberg-rust` than having different defaults distributed over multiple query engines. In my latest commit I changed `set_branch_snapshot` to create branches when they don't exist - just like java. Change is here:
https://github.com/apache/iceberg-rust/pull/587/commits/72b5c0ccdd50625a3049e08148a39b63038a89da

`append_snapshot().set_ref()` is still available for more fine-grained control.